sdk/python/lib/pulumi/resource.py

```diff
@@ -806,10 +806,7 @@ def __init__(self,
             else:
                 # If a provider was specified, add it to the providers map under this type's package
                 # so that any children of this resource inherit its provider.
-                type_components = t.split(":")
-                if len(type_components) == 3:
-                    [pkg, _, _] = type_components
-                    self._providers = {**self._providers, pkg: provider}
+                self._providers = {**self._providers, provider.package: provider}
```

Comment:
Aha so provider gets indexed by provider.package (the one it's for), not the <type> package.